File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/mailbox/disambiguate/DisambiguateMailFormViewModel.kt
@file:Suppress("MagicNumber", "PackageNaming", "LongParameterList")

package app.pantopus.android.ui.screens.mailbox.disambiguate

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.mailbox.v2.ResolveRoutingRequest
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.mailbox.MailboxRepository
import app.pantopus.android.ui.components.IdentityPillar
import app.pantopus.android.ui.screens.homes.invite_owner.ToastPayload
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/** Nav-arg keys for the disambiguate form. */
const val DISAMBIGUATE_MAIL_ID_KEY = "mailId"
const val DISAMBIGUATE_OCR_TEXT_KEY = "ocrText"
const val DISAMBIGUATE_CONFIDENCE_KEY = "confidence"
const val DISAMBIGUATE_ENVELOPE_URL_KEY = "envelopeUrl"

/** One of the three drawer destinations the backend accepts. */
enum class MailRecipientChoice(
    val drawer: String,
    val title: String,
    val subtitle: String,
    val identity: IdentityPillar,
) {
    Personal(
        drawer = "personal",
        title = "Just for me",
        subtitle = "Routes to your personal drawer",
        identity = IdentityPillar.Personal,
    ),
    Home(
        drawer = "home",
        title = "My home household",
        subtitle = "Routes to the shared home drawer",
        identity = IdentityPillar.Home,
    ),
    Business(
        drawer = "business",
        title = "My business inbox",
        subtitle = "Routes to the business team drawer",
        identity = IdentityPillar.Business,
    ),
}

/** Aggregate UI state for the Disambiguate form. */
data class DisambiguateUiState(
    val ocrRecipient: String = "",
    val confidence: Double = 0.0,
    val envelopeUrl: String? = null,
    val selected: MailRecipientChoice? = null,
    val aliasNotes: String = "",
    val isSubmitting: Boolean = false,
    val toast: ToastPayload? = null,
    val shouldDismiss: Boolean = false,
) {
    /** Char-count error when the alias would exceed the schema limit. */
    val aliasError: String?
        get() = if (aliasNotes.trim().length > 255) {
            "Notes must be 255 characters or fewer."
        } else {
            null
        }

    val canSubmit: Boolean get() = selected != null && aliasError == null && !isSubmitting
}

/**
 * POSTs `/api/mailbox/v2/resolve`
 * (`backend/routes/mailboxV2.js:555`).
 *
 * Backend accepts a `drawer` enum (`personal | home | business`) plus
 * an optional alias. The Form design drew "candidate recipients" with
 * avatars; we surface the three drawers as the three rows. See PR
 * description for the discrepancy.
 */
@HiltViewModel
class DisambiguateMailFormViewModel
    @Inject
    constructor(
        private val repo: MailboxRepository,
        savedStateHandle: SavedStateHandle,
    ) : ViewModel() {
        private val mailId: String =
            requireNotNull(savedStateHandle[DISAMBIGUATE_MAIL_ID_KEY]) {
                "DisambiguateMailFormViewModel requires a '$DISAMBIGUATE_MAIL_ID_KEY' nav arg."
            }

        private val _state =
            MutableStateFlow(
                DisambiguateUiState(
                    ocrRecipient = savedStateHandle[DISAMBIGUATE_OCR_TEXT_KEY] ?: "",
                    confidence = savedStateHandle[DISAMBIGUATE_CONFIDENCE_KEY] ?: 0.0,
                    envelopeUrl = savedStateHandle[DISAMBIGUATE_ENVELOPE_URL_KEY],
                ),
            )
        val state: StateFlow<DisambiguateUiState> = _state.asStateFlow()

        fun select(choice: MailRecipientChoice) {
            _state.update { it.copy(selected = choice) }
        }

        fun setAliasNotes(value: String) {
            _state.update { it.copy(aliasNotes = value) }
        }

        fun dismissToast() {
            _state.update { it.copy(toast = null) }
        }

        fun acknowledgeDismiss() {
            _state.update { it.copy(shouldDismiss = false) }
        }

        fun submit() {
            val current = _state.value
            val choice = current.selected ?: run {
                _state.update {
                    it.copy(toast = ToastPayload("Pick a destination first.", isError = true))
                }
                return
            }
            current.aliasError?.let { msg ->
                _state.update { it.copy(toast = ToastPayload(msg, isError = true)) }
                return
            }
            _state.update { it.copy(isSubmitting = true) }
            val trimmedAlias = current.aliasNotes.trim()
            val request =
                ResolveRoutingRequest(
                    mailId = mailId,
                    drawer = choice.drawer,
                    addAlias = if (trimmedAlias.isEmpty()) null else true,
                    aliasString = trimmedAlias.ifEmpty { null },
                )
            viewModelScope.launch {
                when (val result = repo.resolve(request)) {
                    is NetworkResult.Success ->
                        _state.update {
                            it.copy(
                                isSubmitting = false,
                                toast =
                                    ToastPayload(
                                        "Mail routed to ${choice.title.lowercase()}.",
                                        isError = false,
                                    ),
                                shouldDismiss = true,
                            )
                        }
                    is NetworkResult.Failure ->
                        _state.update {
                            it.copy(
                                isSubmitting = false,
                                toast =
                                    ToastPayload(
                                        result.error.message ?: "Couldn't route this mail.",
                                        isError = true,
                                    ),
                            )
                        }
                }
            }
        }
    }

File: frontend/apps/ios/Pantopus/Features/Mailbox/Disambiguate/DisambiguateMailFormViewModel.swift
//
//  DisambiguateMailFormViewModel.swift
//  Pantopus
//
//  POSTs `/api/mailbox/v2/resolve`
//  (`backend/routes/mailboxV2.js:555`). Schema:
//  `resolveRoutingSchema` (line 12) — mailId + drawer (`personal` |
//  `home` | `business`) + optional alias.
//
//  The Form design drew "Possible recipients" as a list of people with
//  avatars. The backend has no candidates endpoint and resolves into
//  one of three drawers. The form translates "drawer choice" into the
//  same radio-row UX with three rows; the optional notes field maps
//  to `aliasString` (when non-empty, sends `addAlias: true`).
//

import Foundation
import Observation

/// One of the three drawer options surfaced as a radio row.
public enum MailRecipientChoice: String, CaseIterable, Identifiable, Sendable {
    case personal
    case home
    case business

    public var id: String { rawValue }

    /// Human-readable row label.
    public var title: String {
        switch self {
        case .personal: "Just for me"
        case .home: "My home household"
        case .business: "My business inbox"
        }
    }

    /// Subtitle copy underneath the row title.
    public var subtitle: String {
        switch self {
        case .personal: "Routes to your personal drawer"
        case .home: "Routes to the shared home drawer"
        case .business: "Routes to the business team drawer"
        }
    }

    /// Identity-pillar tint used by the avatar / accent.
    public var identity: IdentityPillar {
        switch self {
        case .personal: .personal
        case .home: .home
        case .business: .business
        }
    }
}

/// Render state for the Disambiguate form.
public enum DisambiguateMailFormState: Sendable, Equatable {
    case editing
    case error(String)
}

/// ViewModel backing `DisambiguateMailFormView`.
@Observable
@MainActor
final class DisambiguateMailFormViewModel {
    private(set) var state: DisambiguateMailFormState = .editing

    /// Currently-selected recipient row.
    var selectedChoice: MailRecipientChoice?

    /// Optional alias / notes the user types under "Anything else?".
    /// When non-empty, sent as `addAlias: true, aliasString: <value>`.
    var aliasNotes: String = ""

    /// OCR'd recipient text rendered in the envelope card. Wired in by
    /// the screen from the upstream mail item.
    var ocrRecipient: String

    /// AI confidence in the OCR / routing — drives the confidence pill
    /// color and label.
    var confidence: Double

    /// Scanned-envelope image URL (nil → fall through to a placeholder).
    var envelopeImageURL: URL?

    /// Busy flag for the sticky CTA.
    private(set) var isSubmitting: Bool = false

    var toast: ToastMessage?
    /// Set true after a successful resolve so the screen can pop.
    private(set) var shouldDismiss: Bool = false

    /// True while the alias remains over the 255-char schema limit.
    var aliasError: String? {
        let trimmed = aliasNotes.trimmingCharacters(in: .whitespaces)
        if trimmed.count > 255 { return "Notes must be 255 characters or fewer." }
        return nil
    }

    /// Computed "primary action enabled" — drives the sticky CTA.
    var canSubmit: Bool {
        selectedChoice != nil && aliasError == nil && !isSubmitting
    }

    private let mailId: String
    private let api: APIClient

    init(
        mailId: String,
        ocrRecipient: String = "",
        confidence: Double = 0.0,
        envelopeImageURL: URL? = nil,
        api: APIClient = .shared
    ) {
        self.mailId = mailId
        self.ocrRecipient = ocrRecipient
        self.confidence = confidence
        self.envelopeImageURL = envelopeImageURL
        self.api = api
    }

    func select(_ choice: MailRecipientChoice) {
        selectedChoice = choice
    }

    @discardableResult
    func submit() async -> Bool {
        guard let choice = selectedChoice else {
            toast = ToastMessage(text: "Pick a destination first.", kind: .error)
            return false
        }
        if let aliasError {
            toast = ToastMessage(text: aliasError, kind: .error)
            return false
        }
        if !NetworkMonitor.shared.isOnline {
            toast = ToastMessage(
                text: "You're offline. Try again when you're back online.",
                kind: .error
            )
            return false
        }
        isSubmitting = true
        defer { isSubmitting = false }
        let trimmedAlias = aliasNotes.trimmingCharacters(in: .whitespacesAndNewlines)
        let request = ResolveRoutingRequest(
            mailId: mailId,
            drawer: choice.rawValue,
            addAlias: trimmedAlias.isEmpty ? nil : true,
            aliasString: trimmedAlias.isEmpty ? nil : trimmedAlias
        )
        do {
            let response: ResolveRoutingResponse = try await api.request(
                MailboxV2Endpoints.resolve(request)
            )
            toast = ToastMessage(
                text: "Mail routed to \(choice.title.lowercased()).",
                kind: .success
            )
            // Use the backend-confirmed drawer in case the response
            // diverges from what we sent.
            _ = response.drawer
            shouldDismiss = true
            return true
        } catch {
            toast = ToastMessage(
                text: (error as? APIError)?.errorDescription ?? "Couldn't route this mail.",
                kind: .error
            )
            return false
        }
    }

    func acknowledgeDismiss() {
        shouldDismiss = false
    }
}
